docs: migrate 'migrate between bases' (#5409)
canonical/snapcraft: docs/how-to/change-bases/index.rst

keyword is updated:

.. code-block:: diff

    - base: core18
    + base: core20

but further changes will most likely be needed. These will depend on the
original base and the packages that are bundled with the application. This guide
will describe the most common required changes.

Update from an old or absent base
---------------------------------

Migrating a snap from having no base, or ``base: core``, to ``core18`` or ``core20``,
for example, is a more involved process than going from ``core18`` to ``core20``.

This is because building a snap with an old base causes Snapcraft to operate in
compatibility mode.

Compatibility mode is essentially an older (2.43-era) version of Snapcraft, and will
lose the functionality of newer releases.

Package names
-------------

The ``build-packages`` and ``stage-packages`` keys in a snap's project file specify
which packages need to be incorporated during the build and stage parts of the
:external+craft-parts:ref:`parts lifecycle <lifecycle>` and described in build and
staging dependencies.

When no base or core is specified, packages from the Ubuntu 16.04 LTS archive are used

> at build and stage time. The core18 base will use packages from the Ubuntu 18.04 LTS

Reviewer: nit: core, core18, and core20 are sometimes monospaced and sometimes regular text throughoug this doc

Author: Thanks for pointing this out. I thought I caught all of these.

https://github.com/canonical/snapcraft/pull/5409/commits/07ca1eba7f25e9c5e4920fba72f00940e697921d

      parts:
        tarball:
    -     after: [desktop-gtk3]
      apps:
        dwarffortress:
    -     command: desktop-launch $SNAP/wrapper.sh
    +     extensions: [gnome-3-38]
    +     command: wrapper.sh

Audio interfaces
----------------

For applications which play or record audio, the interface names have changed.
Previously the :literalref:`pulseaudio <https://snapcraft.io/docs/pulseaudio-interface>`
interface was used for both playback and recording of audio. This has been replaced by
:literalref:`audio-playback<https://snapcraft.io/docs/audio-playback-interface>` and
``audio-record``:

.. code-block:: diff
    :caption: snapcraft.yaml of Xonotic

      apps:
        xonotic:
          plugs:
    -       pulseaudio
    +       audio-playback

Note that to ensure privacy, ``audio-playback`` is automatically connected but
``audio-record`` isn't.

Application publishers who believe ``audio-record`` should be automatically connected on
install (such as for an audio recording application) should start a thread in the
`store-requests <https://forum.snapcraft.io/c/store-requests>`_ category on the
Snapcraft forum asking for it.

Version scripts
---------------

> The top-level ``version-script`` key has been deprecated in favor of ``adopt-info``.

Author: Nope, I must've missed this one somehow. Thanks for catching that!

https://github.com/canonical/snapcraft/pull/5409/commits/7b0cc7e9d3026cecf8abd6ac8173e935f1a7acc4

Change bases
============

> A :ref:`base snap <base-snaps>` is a special kind of snap that provides a run-time

Reviewer 2: ```suggestion
A :ref:`base snap <base-snaps>` is a special kind of snap that provides a runtime
```

Author: https://github.com/canonical/snapcraft/pull/5409/commits/4f65784eca2e587e23eef8fce54be68394c837a9

environment with a minimal set of libraries common to most applications. They're
transparent to users, but they need to be carefully considered when building a snap.

Each base snap is built from a :ref:`corresponding Ubuntu LTS <reference-bases>`
release. Migrating a snap from one base to the next gives the snap access to newer
packages, extended support, and the latest Snapcraft features, including plugins and
:ref:`extensions <extensions>`.

The complexity of the migration process is directly linked to both dependencies in the
snap's project file and the base snap versions being migrated between.

> At its simplest, migrating from one base snap to another requires only that the base
> keyword is updated:

Reviewer 2: ```suggestion
At its simplest, migrating from one base snap to another requires only that the ``base``
key is updated:
```

Author: https://github.com/canonical/snapcraft/pull/5409/commits/4f65784eca2e587e23eef8fce54be68394c837a9

.. code-block:: diff

    - base: core18
    + base: core20

> but further changes will most likely be needed. These will depend on the

Reviewer 2: ```suggestion
But further changes will most likely be needed. These will depend on the
```

Author: https://github.com/canonical/snapcraft/pull/5409/commits/4f65784eca2e587e23eef8fce54be68394c837a9

original base and the packages that are bundled with the application.

Update from an old or absent base
---------------------------------

Migrating a snap from having no base, or ``base: core``, to core18 or core20,
for example, is a more involved process than going from core18 to core20.

This is because building a snap with an old base causes Snapcraft to operate in
compatibility mode.

Compatibility mode is essentially an older (2.43-era) version of Snapcraft, and will
lose the functionality of newer releases.

Core migration guides
---------------------

Our core migration how-to guides provide detailed instructions and examples to help you

> transition your snap between specific bases.

Reviewer 2: Great!

Author: https://github.com/canonical/snapcraft/pull/5409/commits/4f65784eca2e587e23eef8fce54be68394c837a9
